chronos/internal/asyncengine.nim

```diff
@@ -10,6 +10,94 @@
 
 {.push raises: [].}
 
+## This module implements the core asynchronous engine / dispatcher.
+##
+## The dispatcher acts as a kind of loop that wires events coming from the
+## operating system to tasks scheduled by the program. You must call
+## ``poll`` on it (or a function which does so for you such as ``waitFor`` or
+## ``runForever``) in order to poll for any outstanding events.
+##
+## The underlying implementation is based on epoll on Linux, IO Completion Ports
+## on Windows and select on other operating systems.
+##
+## The ``poll`` function will not, on its own, return any events. Instead
+## an appropriate ``Future`` object will be completed. A ``Future`` is a
+## type which holds a value which is not yet available, but which *may* be
+## available in the future. You can check whether a future is finished
+## by using the ``finished`` function. When a future is finished it means that
+## either the value that it holds is now available or it holds an error instead.
+## The latter situation occurs when the operation to complete a future fails
+## with an exception. You can distinguish between the two situations with the
+## ``failed`` function.
+##
+## Future objects can also store callback procedures which will be called
+## automatically once the future completes.
+##
+## Futures therefore can be thought of as an implementation of the proactor
+## pattern. In this
+## pattern you make a request for an action, and once that action is fulfilled
+## a future is completed with the result of that action. Requests can be
+## made by calling the appropriate functions. For example: calling the ``recv``
+## function will create a request for some data to be read from a socket. The
+## future which the ``recv`` function returns will then complete once the
+## requested amount of data is read **or** an exception occurs.
+##
+## Code to read some data from a socket may look something like this:
+##
+##   .. code-block::nim
+##      var future = socket.recv(100)
+##      future.addCallback(
+##        proc () =
+##          echo(future.read)
+##      )
+##
+## All asynchronous functions returning a ``Future`` will not block. They
+## will not however return immediately. An asynchronous function will have
+## code which will be executed before an asynchronous request is made, in most
+## cases this code sets up the request.
+##
+## In the above example, the ``recv`` function will return a brand new
+## ``Future`` instance once the request for data to be read from the socket
+## is made. This ``Future`` instance will complete once the requested amount
```

Comment:
Its not correct assumptions because as i already said it depends on OS. For example when you perform `recv` you did not send any requests you first subscribe to OS events and perform waiting for the `data_available` event.

Reply:
81a6584a44c69c03c47117c848dbbe63fac79fb5